src/QsCompiler/Transformations/FunctorGeneration.cs

```diff
@@ -148,6 +152,125 @@ public override QsStatementKind OnConjugation(QsConjugation stm)
         }
     }
 
+    /// <summary>
+    /// Scope transformation that splits any nested operation calls into separate statements so
+    /// that they can be properly reversed. This is necessary to avoid out of order execution of the
+    /// automatically generated adjoint. It is safe to do because an adjointable operation must return
+    /// Unit, so any nested calls can be replaced by Unit and those calls moved to separate,
+    /// ordered statements.
+    /// </summary>
+    public class ExtractNestedOperationCalls
+    : SyntaxTreeTransformation<ExtractNestedOperationCalls.TransformationsState>
+    {
+        public class TransformationsState
+        {
+            /// <summary>
+            /// Accumulates statements that have been lifted from the current statement.
+            /// </summary>
+            public List<QsStatement> AdditionalStatements = new List<QsStatement>();
+
+            public Stack<TypedExpression> CurrentExpression = new Stack<TypedExpression>();
+        }
+
+        public ExtractNestedOperationCalls()
+        : base(new TransformationsState())
+        {
+            this.Statements = new ExtractNestedOperationCalls.StatementTransformation(this);
+            this.Expressions = new ExtractNestedOperationCalls.ExpressionTransformation(this);
+            this.ExpressionKinds = new ExtractNestedOperationCalls.ExpressionKindTransformation(this);
+        }
+
+        public class StatementTransformation
```

Comment:
This goes for the expression kind and expression transformation classes as well.

Reply:
I can make these private and also make the `ExtractNestedOperationCalls` that wraps them internal. In fact, all the top level transformations defined in this file can be changed to internal.

Reply:
If the transformation as a whole is internal, than the TransformationState class can also be internal, but ideally there would only be one way to use the transformation and we could limit the use to just that one way with a static wrapper class. Please see my other comment for that.

Reply:
Also I'm not sure about changing the accessibility of the other existing transformations in this file. Ideally they would have been made with more restrictive access initially, but now that they have been released as public others may have taken dependencies on them. There may also be some other reasons why they are public that I just don't know about. At any rate, if you remove things from the existing public API, please make sure to get @bettinaheim 's approval first.

Reply:
Having another static wrapper class would be redundant, since there already exists one in CodeTransformations.cs, and the `GenerateAdjoint` static method is the only place this transformation is used. Unfortunately I can't make it all the way private because they are in different namespaces. Once these classes are internal, they are not in the public API anymore, so it is really that much different to have them be private if it requires adding even more indirection on top of what already exists?

Reply:
The difference between internal vs private is less of an issue here I think because then it scopes any breaking changes to just breaking our own code, which we can manage. Having a single static method as an entry point prevents our own code from jumping into the middle of the transformation, which could be important here as this is a stateful transformation. For example we could do something like `new ExtractNestedOperationCalls().ExpressionKind.OnOperationCall(method, arg);` which would be confusing because we would not have any `CurrentExpression` to utilize. As an internal transformation though, that would only be on us to make sure not to misuse the transformation.

Reply:
It might also be unnecessary overhead, as you suggest, just to police our own code in this case. It's up to you which way you want to go with it. I'll close this thread out and make a separate thread regarding the access modifiers for existing transformations.

Reply:
@bettinaheim To weigh in on the design considerations here...

One option would be to move the static methods in CodeTransformations.cs like `GenerateAdjoint` into FunctorGeneration.cs so that those entrypoints and the code that implements the transformations can be part of the same static class. Then we'd be able to get closer to the design used in other places without needing an intermediate wrapper class. Would that work?